src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,356 @@
+                                    const char *oauthbearer_config,
+                                    void *opaque) {
+        static const rd_ts_t timeout_s = 20;
+        static const int retry         = 3;
+        static const int retry_ms      = 30 * 1000;
+
+        cJSON *json = NULL;
+        cJSON *parsed_token;
+        rd_http_error_t *herr;
+        char *jwt_token;
+        struct curl_slist *headers = NULL;
+        char *post_fields;
+        size_t post_fields_size;
+        char *errstr;
+        const char *token_url;
+
+        char set_token_errstr[512];
+
+        const char *extension_str;
+        size_t extension_cnt;
+        char **extension;
+        char **extension_key_value;
+        size_t extension_key_value_cnt;
+
+        if (rd_kafka_terminating(rk))
+                return;
+
+        errstr = rd_kafka_oidc_build_headers(
+            rk->rk_conf.sasl.oauthbearer.client_id,
+            rk->rk_conf.sasl.oauthbearer.client_secret, &headers);
+        if (errstr != NULL) {
+                rd_kafka_set_fatal_error(rk, RD_KAFKA_RESP_ERR__AUTHENTICATION,
+                                         "Failed to build OAUTHBEARER "
+                                         "OIDC headers: %s",
+                                         errstr);
+                return;
+        }
+
+        /* Build post fields */
+        post_fields_size = strlen("grant_type=client_credentials&scope=") +
+                           strlen(rk->rk_conf.sasl.oauthbearer.scope) + 1;
+        post_fields = rd_malloc(post_fields_size);
+        rd_snprintf(post_fields, post_fields_size,
+                    "grant_type=client_credentials&scope=%s",
+                    rk->rk_conf.sasl.oauthbearer.scope);
+
+        token_url = rk->rk_conf.sasl.oauthbearer.token_endpoint_url;
+
+        herr = rd_http_post_expect_json(rk, token_url, headers, post_fields,
+                                        post_fields_size, timeout_s, retry,
+                                        retry_ms, &json);
+
+        if (unlikely(herr != NULL)) {
+                rd_kafka_log(rk, LOG_ERR, "OIDC",
+                             "Failed to retrieve OIDC "
+                             "token from \"%s\": %s (%d)",
+                             token_url, herr->errstr, herr->code);
+                rd_kafka_oauthbearer_set_token_failure(rk, herr->errstr);
+                rd_http_error_destroy(herr);
+                goto done;
+        }
+
+        parsed_token = cJSON_GetObjectItem(json, "access_token");
+
+        jwt_token     = parsed_token->valuestring;
+        extension_str = rk->rk_conf.sasl.oauthbearer.extensions_str;
+        extension =
+            rd_string_split(extension_str, ',', rd_true, &extension_cnt);
+
+        extension_key_value = rd_kafka_conf_kv_split(
+            (const char **)extension, extension_cnt, &extension_key_value_cnt);
+
+        if (rd_kafka_oauthbearer_set_token(
+                rk, jwt_token, rd_uclock(), "",
```

Comment:
The token expires is actually in the JWT itself. The broker doesn't use the expires here actually.

Reply:
If you look at the set_token() implementation there's two things that makes the current time problematic:
 1. it checks that the lifetime is in the future, which will cause your set_token() calls to fail (was this observed? if not, make sure the lifetime is sent as milliseconds since epoch, not microseconds)
 2. the refresh interval timer is scheduled for 80% of the lifetime.

Reply:
-> "was this observed?" 
I didn't see any error caused by this before. 

Thanks, I will update this part.